explorer/interpreter/value_transform.h

```diff
@@ -0,0 +1,151 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_EXPLORER_INTERPRETER_VALUE_TRANSFORM_H_
+#define CARBON_EXPLORER_INTERPRETER_VALUE_TRANSFORM_H_
+
+#include "explorer/interpreter/value.h"
+
+namespace Carbon {
+
+// A no-op visitor used to implement `IsRecursivelyTransformable`. The
+// `operator()` function returns `true_type` if it's called with arguments that
+// can be used to construct `T`, and `false_type` otherwise.
+template <typename T>
+struct IsRecursivelyTransformableVisitor {
+  template <typename... Args>
+  auto operator()(Args&&... args)
+      -> std::integral_constant<bool, std::is_constructible_v<T, Args...>>;
+};
+
+// A type trait that indicates whether `T` is transformable. A transformable
+// type provides a function
+//
+// template<typename F> void Decompose(F f) const;
+//
+// that takes a callable `f` and passes it an argument list that can be passed
+// to the constructor of `T` to create an equivalent value.
+template <typename T, typename = std::true_type>
+constexpr bool IsRecursivelyTransformable = false;
+template <typename T>
+constexpr bool IsRecursivelyTransformable<
+    T, decltype(std::declval<const T>().Decompose(
+           IsRecursivelyTransformableVisitor<T>{}))> = true;
+
+// Base class for transforms of visitable data types.
+template <typename Derived>
+class TransformBase {
```

Comment:
Given the abstraction here, do you expect to soon be using this for other transforms?